File: top-interview-questions-medium/Linked List/Trees and Graphs/binary-tree-zigzag-level-order-traversal/103.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        result = []
        self.helper(root, 0, result)
        return result

    def helper(self, node, level, res):
        if not node: return
        if len(res) <= level: res.append([])
        if level % 2:
            res[level].insert(0, node.val)
        else:
            res[level].append(node.val)
        self.helper(node.left, level + 1, res)
        self.helper(node.right, level + 1, res)

File: top-interview-questions-medium/Trees and Graphs/binary-tree-zigzag-level-order-traversal/103.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
  const result = []
  const helper = function(node, level) {
    if (!node) return
    if (result.length <= level) result.push([])
    level % 2 ? result[level].unshift(node.val) : result[level].push(node.val)
    helper(node.left, level + 1)
    helper(node.right, level + 1)
  }
  helper(root, 0)
  return result
}
